Data/HashMap/Base.hs

```diff
@@ -1398,6 +1406,115 @@ alterFEager f !k m = (<$> f mv) $ \fres ->
+-- another map if the keys are subsets and the corresponding values are smaller:
+--
+-- >>> isSubmapOfBy (⊑) m1 m2 = keys m1 ⊆ keys m2 && and [ v1 ⊑ v2 | (k1,v1) <- toList m1; let v2 = m2 ! k1 ]
+--
+-- This defines a partial order on maps, for which 'unionWith' is the least
+-- upper bound. More specifically, let @(⊑)@ be a partial order on values for
+-- which @(⊔)@ is the least upper bound. Then for all maps @m1@ and @m2@ it
+-- holds @isSubmapOfBy (⊑) m1 (unionWith (⊔) m1 m2)@ and
+-- @isSubmapOfBy (⊑) m2 (unionWith (⊔) m1 m2)@.
+isSubmapOfBy :: (Eq k, Hashable k) => (v1 -> v2 -> Bool) -> HashMap k v1 -> HashMap k v2 -> Bool
+isSubmapOfBy comp = go 0
+  where
+    -- An empty map is always a submap of any other map.
+    go !_ Empty _ = True
+
+    -- If the first map contains only one entry, lookup the key in the second map.
+    go s (Leaf h1 (L k1 v1)) t2 = lookupCont (\_ -> False) (\v2 _ -> comp v1 v2) h1 k1 s t2
+
+    -- In this case we need to check that for each x in ls1, there is a y in ls2
+    -- such that x ⊑ y. This is the worst case complexity-wise since it requires a O(m*n) check.
+    go _ (Collision h1 ls1) (Collision h2 ls2) =
+      h1 == h2 && subsetArray comp ls1 ls2
+
+    -- To check ls1 ⊆ ls2, we only need to check the entries in ls2 with the hash h1.
+    go s t1@(Collision h1 _) (BitmapIndexed b ls2)
+        | b .&. m == 0 = False
+        | otherwise    =
+            go (s+bitsPerSubkey) t1 (A.index ls2 (sparseIndex b m))
+      where m = mask h1 s
+
+    -- Similar to the previous case we need to traverse l2 at the index for the hash h1.
+    go s t1@(Collision h1 _) (Full ls2) =
+      go (s+bitsPerSubkey) t1 (A.index ls2 (index h1 s))
+
+    -- Insert the collision into a bitmap indexed node and recurse.
+    go s t1@(BitmapIndexed {}) t2@(Collision h2 _) =
+      go s t1 (BitmapIndexed (mask h2 s) (A.singleton t2))
+    go s t1@(Full {}) t2@(Collision h2 _) =
+      go s t1 (BitmapIndexed (mask h2 s) (A.singleton t2))
+
+    -- In cases where the first and second map are bitmap indexed or full,
+    -- traverse down the tree at the appropriate indices.
+    go s (BitmapIndexed b1 ls1) (BitmapIndexed b2 ls2) =
+      submapBitmapIndexed (go (s+bitsPerSubkey)) b1 ls1 b2 ls2
+    go s (BitmapIndexed b1 ls1) (Full ls2) =
+      submapBitmapIndexed (go (s+bitsPerSubkey)) b1 ls1 fullNodeMask ls2
+    go s (Full ls1) (Full ls2) =
+      submapBitmapIndexed (go (s+bitsPerSubkey)) fullNodeMask ls1 fullNodeMask ls2
+    go s (Full ls1) (BitmapIndexed b2 ls2) =
+      submapBitmapIndexed (go (s+bitsPerSubkey)) fullNodeMask ls1 b2 ls2
+
+    -- TODO: I'm not sure about these cases and need help. If we cleared up all
+    -- these cases, we can replace them with a catch-all case go _ _ _ = False
+
+    -- If the second map is empty and the first is not, it cannot be a submap.
+    go _ _ Empty = False
+
+    -- A collision, bitmap indexed an full node always contain at least two
```

Comment:
Where is this documented? (And is it checked?)

Reply:
Yes, these are the cases I'm not sure about, hence the TODO above.

The implementation assumes that this is an invariant of hashmaps. Therefore it is not checked.

Reply:
I'm somewhat concerned that e.g. the results of `delete` or `difference` might violate these assumptions, but I don't know for sure.

@treeowl Do you know more about this?

@svenkeidel I think this is related: https://github.com/haskell-unordered-containers/unordered-containers/issues/161#issuecomment-311246147

Reply:
This is a case which reduces the size of a `BitmapIndexed` node to 1: https://github.com/haskell-unordered-containers/unordered-containers/blob/master/Data/HashMap/Base.hs#L1096

I'll implement these cases instead of relying that this invariant holds.

Reply:
It's not inherently a problem to have a `BitmapIndexed` of length 1, is it? Or is it? Let's work this out for sure. I'd prefer not to have to change the `Eq` instance.

Reply:
It is not a problem. But I relied on this assumption in the implementation of `isSubmapOf`. In other words, it is a bug in `isSubmapOf` I still need to fix.

Reply:
But first I want to understand why my QuickCheck tests did not catch this bug.

Reply:
@svenkeidel If you could create a reproducer that violates these assumptions, that would be very useful, and not just for this PR! :)

Reply:
Ah, I see. The invalid assumption you made is that the array in a `BitmapIndexed` node has at least two elements. The invariant that *does* hold (or had better hold, anyway) is that when a `BitMapIndexed` array only has one element, that element is *not* a `Leaf` or a `Collision`. This, along with the ban on internal `Empty` nodes, ensures that map representation is unique up to collision bucket permutation.

Reply:
This is very helpful. We should document this in the code, best at the type definition.

One more question. From studying `delete`, I learned that `Full` nodes are downgraded to `BitmapIndexed` nodes as soon as a child node becomes empty. Similarly, `Collision` nodes are downgraded to `Leafs` when there is only a single key left. I assume this means that `Full` and `Collision` nodes have at least 2 elements. Correct?

Reply:
`Collision` nodes have at least two. `Full` nodes have *exactly* sixteen, where 16 comes about from some fixed definition in the module. (Don't remember what it's called, or exactly how it gives 16.)